[Week4][Davidyang2149] Refactor To-do App by Redux - 구현 완료 (1차 리팩토링 완료)

DavidYang2149 opened the pull request:
# 4강 리덕스 - 4주 차 과제 1
## 기능 구현
- [x] 상태 관리를 Redux로 변경해도 모든 컴포넌트 테스트가 통과해야 합니다. 추가로 reducer들에 대한 테스트를 추가해 주세요.

## 제한 조건
- [ ] 모든 기능은 TDD로 구현합니다. 컴포넌트와 reducer에 대한 테스트가 존재해야 합니다. 먼저 테스트를 추가하고 commit 한 뒤 실제 기능을 추가하고 commit합니다. commit 메시지 로그는 아래와 같이 남아야 합니다.
```sh
Restaurant에 대한 테스트를 추가한다
Restaurant을 구현한다
Restaurant를 리팩토링한다
```
- [x] if는 사용할 수 있어도 else 사용하실 수 없습니다. GuardClauses 방법을 사용해주세요.
- [x] switch는 사용하실 수 없습니다.
- [x] let은 사용하실 수 없습니다. const만을 사용하여 문제를 해결해주세요.
- [x] 함수 이름과 변수 이름에 줄임말은 사용하실 수 없습니다. 길더라도 명확하게 써주세요.
- [x] indent(인덴트, 들여쓰기) depth를 1로 유지해주세요. 예를 들어, for문 안에 if문이 있으면 indent depth는 2입니다. depth를 줄이는 좋은 방법은 함수(또는 메소드)를 분리하면 됩니다.

## 요구사항
- [x] ESLint를 돌린 결과 아무런 문제가 없어야 합니다.
- [x] 모든 인수 테스트를 통과시켜야 합니다.
- [x] 한 파일에는 하나의 컴포넌트만 있어야 합니다.
- [x] 테스트 커버리지 100%를 달성해야 합니다.

## 작업내용 및 질의
- [x] 구현 작업 완료
- [x] 1차 리팩토링 완료

DavidYang2149 committed:
Update package.json

4강 리덕스 - 4주 차 과제 1
- redux, react-redux 설치.

DavidYang2149 committed:
Update App.jsx

4강 테스트 - 4주 차 과제 1
- App.jsx 업데이트 작업.
- 관심사 분리를 위한 코드 작업 진행 중.

DavidYang2149 committed:
Add store.js

4강 리덕스 - 4주 차 과제 1
- store.js 생성 및 초기화.

DavidYang2149 committed:
Update App.jsx and index.jsx

4강 리덕스 - 4주 차 과제 1
- reducer 적용을 위한 세팅(진행 중).

DavidYang2149 committed:
Update App.jsx and store.js

4강 리덕스 - 4주 차 과제 1
- App.js & store.js에 reducer 샘플 만들기.

DavidYang2149 committed:
Update App.jsx and store.js

4강 리덕스 - 4주 차 과제 1
- App.js & store.js에 reducer 분리 작업 완료.

DavidYang2149 committed:
Refactor App.js Add actions.js for actions create

4강 리덕스 - 4주 차 과제 1
- App.js : actions 분리 작업.
- actions.js 생성 : actions 생성 작업.

DavidYang2149 committed:
Add __mocks__/react-redux.js

4강 리덕스 - 4주 차 과제 1
- __mocks__/react-reudx.js 생성.
- 테스트 코드 작성을 위한 목업 작성.

DavidYang2149 committed:
Refactor App.test.jsx

4강 리덕스 - 4주 차 과제 1
- App.test.jsx 리덕스 테스트를 위한 테스트 코드 리팩토링.

DavidYang2149 committed:
Refactor reducer.js, reducer.test.js, store.js for TDD

4강 리덕스 - 4주 차 과제 1
- reducer의 테스트 코드 작성을 위한 리팩토링(진행 중).

DavidYang2149 committed:
Refactor reducer.js, reducer.test.js Update func

4강 리덕스 - 4주 차 과제 1
- reducer.text.js 테스트 과정에서 기능 오류 발견에 따른 원본 수정.

DavidYang2149 committed:
Refactor reducer.test.js for TDD

4강 리덕스 - 4주 차 과제 1
- reducer의 테스트 코드 작성을 위한 리팩토링(App.jsx용 전체 완료).

DavidYang2149 committed:
Refactor reudcer.test.js and Add ListContainer.jsx

4강 리덕스 - 4주 차 과제 1
- reducer.test.js 리팩토링.
- ListContainer.jsx 추가.

Kihyun92 commented on src/index.jsx:
```diff
@@ -1,9 +1,17 @@
+
 ReactDOM.render(
-  <App />,
+  (
+    <Provider store={store}>
+      <App />
+    </Provider>
+  ),
```
```suggestion
    <Provider store={store}>
      <App />
    </Provider>,
```
괄호 없이 작성해도 될 것 같아요~

DavidYang2149 replied:
바로 이해가 되지 않아 구현단계에서는 강의를 따라하는 것에 초점을 두었습니다.
구현 완료 후 리팩토링에 들어가도록 하겠습니다!

괄호가 없으니 더 깔끔해보이네요!

Kihyun92 commented on src/reducer.js:
```diff
@@ -0,0 +1,40 @@
+};
+
+export default function reducer(state = initialState, action) {
+  if (action.type === 'updateTaskTitle') {
+    return {
+      ...state,
+      taskTitle: action.payload.taskTitle,
+    };
+  }
+
+  if (action.type === 'addTask') {
+    const { newId, taskTitle, tasks } = state;
+
+    if (!taskTitle) {
+      return state;
+    }
+
+    return {
+      ...state,
+      newId: newId + 1,
+      taskTitle: '',
+      tasks: [...tasks, { id: newId, title: taskTitle }],
+    };
+  }
+
+  if (action.type === 'deleteTask') {
+    const { tasks } = state;
+
+    return {
+      ...state,
+      tasks: tasks.filter((task) => task.id !== action.payload.id),
+    };
+  }
```
`if`문을 쓰지 않고 작성할 방법이 있을까요?? 🤔

DavidYang2149 replied:
object를 이용하면 가능할 것 같은데, 좀 더 고민해보겠습니다!

Kihyun92 replied:
여러 방법이 있겠지만, [Object Literal lookups](https://ultimatecourses.com/blog/deprecating-the-switch-statement-for-object-literals) 을 참고해보셔도 좋을 것 같아요~ 물론 다른 방법도 좋구요 :)

Kihyun92 commented on src/reducer.test.js:
```diff
@@ -0,0 +1,78 @@
+      });
+    });
+
+    context('without task title', () => {
+      it('doesn\'t work', () => {
```
`with` <-> `without`으로 서로 대비되는 테스트임을 나타내는 것도 좋지만, 조금 더 이 테스트에 대해 명확하게 표현해줘도 좋을 것 같아요~

DavidYang2149 replied:
no나 not을 적용하는 것으로 수정해보았습니다.

Kihyun92 commented on src/reducer.test.js:
```diff
@@ -0,0 +1,78 @@
+      });
+    });
+
+    context('without existed task ID', () => {
+      it('remove a task from tasks', () => {
```
마찬가지로 context내의 문장이 `with no existed task ID` 혹은 `with not existed task ID`와 같다면 어떤가요? 그리고 it 내의 문장 또한 해당 테스트를 조금 더 명확히 나타내게 할 수 없을까요?

DavidYang2149 replied:
생각해보니 task ID 가 존재하면 remove 작업도 진행되지 않기 때문에 dosen't work 문구를 추가했습니다!

Kihyun92 reviewed (changes requested):
좋습니다~ 강의의 설명에 맞게 코드를 다뤄보는 것도 좋지만, 더 리펙토링 할 순 없는지 고민하면서 해보시면 더 기억에 남을 것 같습니다~

DavidYang2149 committed:
Add ListContainer.test.jsx for divide Page

4강 리덕스 - 4주 차 과제 1
- Page 리팩토링을 위한 ListContainer.test.jsx 추가.

DavidYang2149 committed:
Add InputContainer.js and InputContainer.text.js 추가

4강 리덕스 - 4주 차 과제 1
- InputContainer.js 추가.
- InputContainer.test.js 추가.

DavidYang2149 committed:
Refactor Page.jsx and Page.test.jsx

4강 리덕스 - 4주 차 과제 1
- Page.jsx 리팩토링 - InputContainer.jsx 분리.
- Page.test.jsx 리팩토링 - Inputcontainer.test.jsx 분리.

DavidYang2149 committed:
Refactor App.jsx and Page.jsx

4강 리덕스 - 4주 차 과제 1
- App.jsx 리팩토링 - App과 Page의 역할 중복에 따른 컨테이너 이전.
- Page.jsx 리팩토링 - App과 Page의 역할 중복에 따른 삭제.

DavidYang2149 committed:
Refactor index.jsx

4강 리덕스 - 4주 차 과제 1
- index.jsx 리팩토링 - 소괄호 제거.

DavidYang2149 committed:
Refactor InputContainer.test.jsx

4강 리덕스 - 4주 차 과제 1
- test coverage 누락 테스트 추가.
- 중복값 const 변수로 생성.

DavidYang2149 committed:
Refactor ListContainer.test.jsx

4강 리덕스 - 4주 차 과제 1
- ListContainer.test.jsx 리팩토링 - 테스트 커버리지 누락 테스트 추가.

DavidYang2149 committed:
Refactor reducer.test.js

4강 리덕스 - 4주 차 과제 1
- reducer.test.js 리팩토링 - 테스트 커버리지 추가(진행 중).

DavidYang2149 committed:
Refactor reducer.test.js

4강 리덕스 - 4주 차 과제 1
- reducer.test.js 리팩토링 - 테스트 문구 수정.

Kihyun92 commented on src/reducer.test.js:
```diff
@@ -0,0 +1,89 @@
+    });
+  });
+
+  describe('none of all', () => {
+    const state = reducer({
+      undefined,
+    }, { type: undefined });
+
+    it('return state and nothing change', () => {
+      expect(state).toBe(state);
+    });
```
`none of all`이라는 문장은 이 테스트를 설명하는데 충분하지 않은 것 같아요! 조금 더 명확히 의도를 드러내도록 수정해보면 좋겠습니다.

그리고 부족한 coverage는 이 테스트가 이유일 것 같습니다. `type`만 세팅되어있는 것 같은데, `payload` 정보도 넘겨줘야 하지 않을까요?

그리고 이렇게까지 복잡하게 하지 않으려면 어떻게 하면 좋을까요? 한번 고민해보시면 좋을 것 같아요 :)

DavidYang2149 replied:
말씀해주신 내용을 토대로 수정하였습니다. 감사합니다!

Kihyun92 reviewed (changes requested):
음 터미널에 출력되는 로그를 확인해봐야하겠지만, `reducer.js`에서 `action`이 없는 경우에 대한 처리가 되어있나요?

DavidYang2149 committed:
Refactor reducer.test.js

4강 리덕스 - 4주 차 과제 1
- reducer.test.js 리팩토링 - 테스트 문구 수정.
- reducer.test.js 리팩토링 - 커버리지 누락 건 수정.

DavidYang2149 committed:
Refactor reducer.js

4강 리덕스 - 4주 차 과제 1
- reducer.js 리팩토링 - if문을 object로 수정.

Kihyun92 reviewed (approved):
좋습니다! 과제 2로 빠르게 넘어갑시다~ 👍
